plugins/Login/Controller.php

```diff
@@ -303,14 +303,15 @@ protected function authenticateAndRedirect($login, $password, $urlToRedirect = f
         $this->passwordResetter->removePasswordResetInfo($login);
 
         if (empty($urlToRedirect)) {
-            $referrer = Url::getReferrer();
-            $module = Common::getRequestVar('module', '', 'string');
+            $redirect = Common::unsanitizeInputValue(Common::getRequestVar('form_redirect', false));
+            $redirectParams = UrlHelper::getArrayFromQueryString(UrlHelper::getQueryFromUrl($redirect));
+            $module = Common::getRequestVar('module', '', 'string', $redirectParams);
             // when module is login, we redirect to home...
-            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $referrer) {
-                $host = Url::getHostFromUrl($referrer);
+            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $redirect) {
```

Comment:
Be good to check that `$module` is not empty...

Reply:
may actually not be needed though because we do validate a host is set below just seeing